specgen/specgen/main.go

```diff
@@ -0,0 +1,117 @@
+// Command specgen provides a utility for generating utls.ClientHelloSpecs. This command starts an
+// HTTPS server which responds to every request with a ClientHelloSpec corresponding to the TLS
+// ClientHello the server received. The spec is logged for every request as well. Make a request of
+// the server via web browser to see a ClientHelloSpec for the browser.
+//
+// It is recommended that mkcert be used to generate the keypair for this server. This ensures
+// browsers do not flag and block the page. See https://github.com/FiloSottile/mkcert.
+package main
+
+import (
+	"bytes"
+	"crypto/tls"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/getlantern/utls/specgen"
+	"github.com/getlantern/utls/specgen/specgen/hellocap"
+)
+
+const certFileDefaultText = "pre-defined, self-signed cert"
+
+var (
+	addr          = flag.String("addr", "localhost:0", "address to listen on")
+	certFile      = flag.String("cert", certFileDefaultText, "PEM-encoded TLS certificate file")
+	keyFile       = flag.String("key", "", "PEM-encoded TLS key file")
+	packagePrefix = flag.String("prefix", "", "package prefix; see specgen.WriteHelloSpec")
+)
+
+func handleRequest(rw http.ResponseWriter, req *http.Request) {
+	hello, err := hellocap.FromRequest(req)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintln(rw, "error capturing ClientHello")
+		log.Println("error capturing ClientHello:", err)
+		return
+	}
+
+	buf := new(bytes.Buffer)
+	err = specgen.WriteHelloSpec(buf, hello, *packagePrefix)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintln(rw, "error writing ClientHello")
+		log.Println("error writing ClientHello:", err)
+		return
+	}
+
+	ua := req.Header.Get("User-Agent")
+
+	fmt.Fprintln(rw, buf.String())
+	log.Printf("captured ClientHello from %s:\n%s\n", ua, buf.String())
+}
+
+func main() {
+	flag.Parse()
+
+	var cert tls.Certificate
+	var err error
+	if *certFile == certFileDefaultText {
+		cert, err = tls.X509KeyPair([]byte(defaultCertPEM), []byte(defaultKeyPEM))
+	} else {
+		cert, err = tls.LoadX509KeyPair(*certFile, *keyFile)
+	}
+	if err != nil {
+		log.Panicf("failed to load cert: %v", err)
+	}
+
+	s, err := hellocap.NewServer(http.HandlerFunc(handleRequest), *addr, cert)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	log.Println("Listening for incoming connections on", s.Addr())
+
+	if err := s.ListenAndServe(); err != nil {
+		log.Panic(err)
+	}
+}
+
+// A self-signed TLS key pair valid until 2032.
+var (
```

Comment:
You can generate these, and the code will last longer than 10 years :)

Reply:
Lol, 10 years seems like plenty to me